index_matroska.md

```diff
@@ -136,7 +136,14 @@ As an EBML Document Type, Matroska adds the following constraints to the EBML sp
 
 ## Matroska Design
 
-All top-levels elements (Segment and direct sub-elements) are coded on 4 octets -- i.e. class D elements.
+The Root Element and all Top-Levels Elements use 4 octets for their EBML Element ID -- i.e. Segment and direct children of Segment.
+
+Matroska doesn't allow using EBML Empty Elements, elements present in the file but with a length of zero.
```

Comment:
I'm a bit hesitant to forbid zero-length elements. I'm thinking mostly about user-provided data: chapters & tags. To me it seems quite possible that people might want to include a string-tag but not set a value yet. This wouldn't be allowed with this change.

Reply:
> I'm thinking mostly about user-provided data: chapters & tags. To me it seems quite possible that people might want to include a string-tag but not set a value yet. This wouldn't be allowed with this change.

Very good catch.

Reply:
The wording is wrong, the goal is to forbid empty elements *with default value that is not zero/empty string*. User values like chapter names or tag values don't have a default value (even less non-zero).

Reply:
Actually that's a **SHOULD NOT** and not **MUST NOT** with the accepted cases being described in the next paragraph. It is missing strings though.

Reply:
And we do not have a way to express default values as "" for strings, so strings are excluded from Empty Elements not matter what.

Reply:
I turned this into a more specific MUST NOT

Reply:
> the goal is to forbid empty elements with default value that is not zero/empty string.

I don't really understand why should this be banned.

If a reader reads a zero-length(empty) EBML-Integer element with a default value of 2, the value 2 is used and not the type default value 0.

Reply:
The reason is given in the next sentence

> This is to preserve compatibility with existing parsers that didn't interpret this feature properly

See #500

Reply:
OK. But like you wrote

> This makes some element temporarily invalid.

And I ask me what is worse:
(maybe) a break of old (very old, no longer maintained) parsers
or make existing parameters invalid.

Reply:
It's neither nor.

As of today libebml has no fix (and no PR AFAIK) to fix this issue. It will likely involve some ABI changes that will take more time to implement and propagate on Linux distributions.

The only invalid parameter is `ChapterLanguage`, every thing is or has been cleaned. And for that field players already assume the default value no matter what.

Reply:
> The only invalid parameter is ChapterLanguage,

The `ChapLanguage` element is not a non-mandatory element and so this is not the case for this element.
If the `ChapLanguage` element is not present or written with a zero-length of data then "eng" is used.

Here a list of non-mandatory not deprecated elements with a default value not equal to 0

* `MatrixCoefficients`
* `TransferCharacteristics`
* 'Primaries'
* 'CueBlockNumber'
* 'TargetTypeValue'

In this list is no String-element!

There are many non-mandatory elements with a default value of 0. All this elements must be written, and actually is the element header enough, because there is a default value which can be used if the data size is zero.
But when the zero-length feature is forbidden then the entire element must be written and the "default  atritubute" is more or less useless or can be omit.

> As of today libebml has no fix (and no PR AFAIK) to fix this issue.

Yes that's true, but such a fix is trivial from my humble point of view.
And when I understand the libebml code right then breaks this fix nothing.

Reply:
Added some more details, especially on strings.

* integer/float with default value = 0, Empty Element is allowed
* integer/float with default value != 0, Empty Element is NOT allowed
* string with default value = "", impossible in EBML Schema
* string with default value != "", Empty Element is NOT allowed

Reply:
For `MatrixCoefficients`, `TransferCharacteristics`, `Primaries`, `CueBlockNumber` and `TargetTypeValue`, they are in line with the rule 
> any element that is not mandatory **SHOULD NOT** have a default value

It's a "SHOULD NOT" rule, not a "MUST NOT". Meaning in some case they might have a default value. Nevertheless they won't be using Empty Elements if the spec is followed

> Matroska writers **MUST NOT** use EBML Empty Elements, elements present in the file but with a length of zero,
if the element has a default value that is not 0 for integers/dates and 0x0p+0 for floats.